Use go discover

magiconair opened the pull request:
not yet fully tested and `command/agent.go` needs some more work.

magiconair commented:
This patch replaces the custom discovery code with go-discover and provides a unified interface for configuration by using the `go-discover` `key=val key=val ...` DSL. Instead of adding a new parameter for specifying this the `--retry-join` parameter has been extended to allow a `provider=xxx` field which then triggers a go-discover lookup. I'd like to get some feedback whether this approach is OK or too clever.

slackpad commented on agent/retry_join.go:
```diff
@@ -2,49 +2,45 @@ package agent
+	var disco string
+	for _, addr := range cfg.RetryJoin {
+		if strings.Contains(addr, "provider=") {
+			disco = addr
```
We probably want to throw an error if they do multiple discos so we don't just quietly do last-one-wins, or we could support multiple discos, too, which wouldn't be too hard now.

slackpad replied:
You could even call discover in here and append all the addresses to the address array, which is simpler.

magiconair replied:
I think this could work for multi-cloud setups but messing this up is a serious operator error. I fail to see how someone can shoot themself in the foot by accident. I can simplify the code a bit  more.

slackpad commented on agent/retry_join.go:
```diff
@@ -2,49 +2,45 @@ package agent
+				a.logger.Printf("[ERR] agent: %s", err)
 			}
-			a.logger.Printf("[INFO] agent: Discovered %d servers from Azure", len(servers))
+			a.logger.Printf("[ERR] agent: Discovered servers: %v", servers)
```
This should not be ERR, INFO is good. Could we show just the provider name in the log line for context?

magiconair replied:
yep

slackpad commented on command/agent.go:
```diff
@@ -431,15 +431,15 @@ func (cmd *AgentCommand) readConfig() *agent.Config {
 	// EC2 and GCE discovery are mutually exclusive
-	if cfg.RetryJoinEC2.TagKey != "" && cfg.RetryJoinEC2.TagValue != "" && cfg.RetryJoinGCE.TagValue != "" {
+	if cfg.DeprecatedRetryJoinEC2.TagKey != "" && cfg.DeprecatedRetryJoinEC2.TagValue != "" && cfg.DeprecatedRetryJoinGCE.TagValue != "" {
 		cmd.UI.Error("EC2 and GCE discovery are mutually exclusive. Please provide one or the other.")
```
We should be able to delete this and support multiple. Not that people would do this, but it doesn't need this nasty special case check.

magiconair replied:
yes. see earlier comment.

slackpad commented on command/agent.go:
```diff
@@ -431,15 +431,15 @@ func (cmd *AgentCommand) readConfig() *agent.Config {
-		if cfg.RetryJoinEC2.TagKey == "" || cfg.RetryJoinEC2.TagValue == "" {
+	if cfg.DeprecatedRetryJoinEC2.TagKey != "" || cfg.DeprecatedRetryJoinEC2.TagValue != "" {
+		if cfg.DeprecatedRetryJoinEC2.TagKey == "" || cfg.DeprecatedRetryJoinEC2.TagValue == "" {
 			cmd.UI.Error("tag key and value are both required for EC2 retry-join")
```
This seems like an error that should come out of go-discover now.

magiconair replied:
yes and it does.

magiconair commented on agent/config.go:
```diff
@@ -20,6 +21,7 @@ import (
 	"github.com/hashicorp/consul/tlsutil"
 	"github.com/hashicorp/consul/types"
 	"github.com/hashicorp/consul/watch"
+	discover "github.com/hashicorp/go-discover"
```
goimports automatically creates an alias.

magiconair commented on agent/config.go:
```diff
@@ -1183,6 +1179,52 @@ func DecodeConfig(r io.Reader) (*Config, error) {
+		result.DeprecatedRetryJoinEC2 = RetryJoinEC2{}
+
+		fmt.Fprintf(os.Stderr, "==> DEPRECATION: retry_join_ec2 is deprecated."+
+			"Please add %q to retry_join\n", m)
```
I'll add the `json:"-"` and redact the output. Since `m` is local to the `if` block I can just overwrite the fields before I print the error message.

magiconair commented:
I've addressed the review comments and will update the docs. I ran a test on AWS where the join worked but I leader election failed. Got this output. Not sure whether this is related to my change. Will continue on Monday.

I'm curious why it logs a "wan" event.

```
./consul agent -server -data-dir data -retry-join 'provider=aws tag_key=consul tag_value=server access_key_id=xxx secret_access_key=xxx'
==> Starting Consul agent...
==> Consul agent running!
           Version: 'v0.9.0-54-ge7fb9e0d-dev (e7fb9e0d)'
           Node ID: '445c5dc0-53a0-6928-39b7-274098b3ca74'
         Node name: 'ip-10-0-1-169'
        Datacenter: 'dc1'
            Server: true (bootstrap: false)
       Client Addr: 127.0.0.1 (HTTP: 8500, HTTPS: -1, DNS: 8600)
      Cluster Addr: 10.0.1.169 (LAN: 8301, WAN: 8302)
    Gossip encrypt: false, RPC-TLS: false, TLS-Incoming: false

==> Log data will now stream in as it occurs:

    2017/07/29 11:02:44 [INFO] raft: Initial configuration (index=0): []
    2017/07/29 11:02:44 [INFO] raft: Node at 10.0.1.169:8300 [Follower] entering Follower state (Leader: "")
    2017/07/29 11:02:44 [INFO] serf: EventMemberJoin: ip-10-0-1-169.dc1 10.0.1.169
    2017/07/29 11:02:44 [INFO] serf: EventMemberJoin: ip-10-0-1-169 10.0.1.169
    2017/07/29 11:02:44 [INFO] agent: Started DNS server 127.0.0.1:8600 (udp)
    2017/07/29 11:02:44 [INFO] consul: Adding LAN server ip-10-0-1-169 (Addr: tcp/10.0.1.169:8300) (DC: dc1)
    2017/07/29 11:02:44 [INFO] consul: Handled member-join event for server "ip-10-0-1-169.dc1" in area "wan"
    2017/07/29 11:02:44 [INFO] agent: Started DNS server 127.0.0.1:8600 (tcp)
    2017/07/29 11:02:44 [INFO] agent: Started HTTP server on 127.0.0.1:8500
    2017/07/29 11:02:44 [INFO] agent: Joining cluster...
    2017/07/29 11:02:44 [INFO] discover-aws: Looking up region
    2017/07/29 11:02:44 [INFO] discover-aws: Region is eu-west-1
    2017/07/29 11:02:44 [INFO] discover-aws: Filter instances by consul=server
    2017/07/29 11:02:44 [INFO] discover-aws: Found 31302e302e312e313331
    2017/07/29 11:02:44 [INFO] discover-aws: Found 31302e302e312e313639
    2017/07/29 11:02:44 [INFO] discover-aws: Found 31302e302e312e313938
    2017/07/29 11:02:44 [ERR] agent: Discovered servers: [10.0.1.131 10.0.1.169 10.0.1.198]
    2017/07/29 11:02:44 [INFO] agent: (LAN) joining: [10.0.1.131 10.0.1.169 10.0.1.198]
    2017/07/29 11:02:44 [INFO] serf: EventMemberJoin: ip-10-0-1-131 10.0.1.131
    2017/07/29 11:02:44 [INFO] serf: EventMemberJoin: ip-10-0-1-198 10.0.1.198
    2017/07/29 11:02:44 [INFO] consul: Adding LAN server ip-10-0-1-131 (Addr: tcp/10.0.1.131:8300) (DC: dc1)
    2017/07/29 11:02:44 [INFO] consul: Adding LAN server ip-10-0-1-198 (Addr: tcp/10.0.1.198:8300) (DC: dc1)
    2017/07/29 11:02:44 [INFO] serf: EventMemberJoin: ip-10-0-1-198.dc1 10.0.1.198
    2017/07/29 11:02:44 [INFO] serf: EventMemberJoin: ip-10-0-1-131.dc1 10.0.1.131
    2017/07/29 11:02:44 [INFO] consul: Handled member-join event for server "ip-10-0-1-198.dc1" in area "wan"
    2017/07/29 11:02:44 [INFO] consul: Handled member-join event for server "ip-10-0-1-131.dc1" in area "wan"
    2017/07/29 11:02:44 [INFO] agent: (LAN) joined: 3 Err: <nil>
    2017/07/29 11:02:44 [INFO] agent: Join completed. Synced with 3 initial agents
    2017/07/29 11:02:51 [ERR] agent: failed to sync remote state: No cluster leader
    2017/07/29 11:02:53 [WARN] raft: no known peers, aborting election
    2017/07/29 11:03:18 [ERR] agent: Coordinate update error: No cluster leader
    2017/07/29 11:03:24 [ERR] agent: failed to sync remote state: No cluster leader
    2017/07/29 11:03:40 [ERR] agent: Coordinate update error: No cluster leader
    2017/07/29 11:03:56 [ERR] agent: failed to sync remote state: No cluster leader
    2017/07/29 11:04:10 [ERR] agent: Coordinate update error: No cluster leader
```

slackpad commented:
> leader election failed

Looks like the join worked fine - you need to add `-bootstrap-expect=3` to your command lines and I think it will work OK.

slackpad commented on agent/config.go:
```diff
@@ -1193,6 +1193,10 @@ func DecodeConfig(r io.Reader) (*Config, error) {
 		result.DeprecatedRetryJoinEC2 = RetryJoinEC2{}
 
+		// redact m before output
+		m["access_key_id"] = "xxx"
```
Let's change `xxx` to `<hidden>` to match what we do when we redact tokens.

magiconair commented:
Why does it say `member join event in area "wan"`?

slackpad commented:
> Why does it say member join event in area "wan"?

That's the join flooder at work. It makes sure that any servers that are known in the LAN are also known in the WAN. Before we did this it was super annoying to remember to join both, so we wrote this little goroutine to keep them in one-way sync - https://github.com/hashicorp/consul/blob/v0.9.0/agent/consul/flood.go

magiconair commented:
It's a bit confusing if there is no WAN config.

slackpad commented:
> It's a bit confusing if there is no WAN config.

Yeah we've had a couple users ask about it, though there's not a great thing to key off of. I suppose we could start flooding only once there's some other thing in the WAN so you don't get this behavior.

magiconair commented:
I've pushed the updated docs but need to read them again in the full context. Also, need to test some of the different auth methods for different providers. So far I've only tested the static configuration.

preetapan commented on website/source/docs/agent/options.html.md:
```diff
@@ -261,9 +344,13 @@ will exit with an error at startup.
 * <a name="_retry_join_azure_tag_value"></a><a href="#_retry_join_azure_tag_value">`-retry-join-azure-tag-value`
   </a> - The Azure instance tag value to filter on.
 
+  As of Consul 0.9.1 this parameter has been deprecated. See [-retry-join](#_retry_join) for details.
```
There should be a comma after `As of Consul 0.9.1`, or you could rewrite the sentence as "This parameter has been deprecated as of Consul 0.9.1"

magiconair commented:
Testing go-discover cmd line on AWS:

- [x] AWS static credentials auth OK
- [x] AWS env vars auth OK
- [x] AWS credentials file in default location OK
- [x] AWS credentials file via `AWS_SHARED_CREDENTIALS_FILE` OK
- [x] AWS EC2 instance IAM role OK

Testing consul with go-discover cmd line on AWS:

- [x] AWS static credentials auth OK
- [x] AWS env vars auth OK
- [x] AWS credentials file in default location OK
- [x] AWS credentials file via `AWS_SHARED_CREDENTIALS_FILE` OK
- [x] AWS EC2 instance IAM role OK

AWS test completed OK

magiconair commented:
Testing go-discover on GCE:

- [X] GCE static credentials file OK
- [X] GCE credentials file path in `GOOGLE_APPLICATION_CREDENTIALS` OK
- [X] GCE credentials file in default location OK
- [X] GCE instance role OK
- [X] GCE discovery with zone pattern regexp OK

Testing consul with go-discover on GCE:

- [X] GCE static credentials file OK
- [X] GCE credentials file path in `GOOGLE_APPLICATION_CREDENTIALS` OK
- [X] GCE credentials file in default location OK
- [X] GCE instance role OK
- [X] GCE discovery with zone pattern regexp OK

GCE test completed OK
